gradle fails to detect Java version as of jdk-10-ea+36

### Expected Behavior
Under some conditions, gradle will attempt to determiner the version of Java by launching 'java -version' and parsing the output. The expected behavior is that this should continue to work.

### Current Behavior
gradle fails to detect the version of java starting with jdk-10-ea+36.

### Context
The JavaFX build uses gradle to produce builds. When using jdk-10+36 as our boot JDK, gradle started failing to detect the java version.

### Steps to Reproduce (for bugs)
1. Unzip the attached test program, which can also be found at:

https://goo.gl/MG84mq

[gradle-java-version.zip](https://github.com/gradle/gradle/files/1572746/gradle-java-version.zip)

2. Download and install jdk-10-ea+36 from here:

http://jdk.java.net/10/


3. Unpack jdk-10 in some directory, and set JAVA_HOME to that directory.



4. Edit the build.gradle script to change the value of the 'JAVA' (and 'JAVAC') var to point to the java executable, but using a different path string than JAVA_HOME. On Linux, the easiest way to do this is to use two forward slashes rather than one somewhere in the path. For example:

def JAVA = "/some/directory/jdk-10//bin/java"


5. Using a recent version of gradle (I used 4.4, which is the latest as of today), run the following:

    gradle build

this will fail with the following stack trace:

Starting process 'command '/localhome/kcr/jdks/jdk-10//bin/java''. Working directory: /localhome/kcr/java/bugs/gradle-java-version Command: /localhome/kcr/jdks/jdk-10//bin/java -version
Successfully started process 'command '/localhome/kcr/jdks/jdk-10//bin/java''
:test FAILED
:test (Thread[Task worker for ':',5,main]) completed. Took 0.069 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> Could not determine Java version using executable /localhome/kcr/jdks/jdk-10//bin/java.

* Try:
Run with --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':test'.
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:50)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.GradleException: Could not determine Java version using executable /localhome/kcr/jdks/jdk-10//bin/java.
	at org.gradle.internal.jvm.inspection.DefaultJvmVersionDetector.parseJavaVersionCommandOutput(DefaultJvmVersionDetector.java:75)
	at org.gradle.internal.jvm.inspection.DefaultJvmVersionDetector.getJavaVersion(DefaultJvmVersionDetector.java:58)
	at org.gradle.internal.jvm.inspection.CachingJvmVersionDetector.getJavaVersion(CachingJvmVersionDetector.java:57)
	at org.gradle.api.tasks.testing.Test.getJavaVersion(Test.java:212)
	at org.gradle.api.tasks.testing.Test_Decorated.getJavaVersion(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.api.internal.project.taskfactory.TaskPropertyInfo$2.create(TaskPropertyInfo.java:89)
	at org.gradle.util.SingleMessageLogger.whileDisabled(SingleMessageLogger.java:230)
	at org.gradle.api.internal.project.taskfactory.TaskPropertyInfo.getValue(TaskPropertyInfo.java:86)
	at org.gradle.api.internal.tasks.TaskPropertyValue.validate(TaskPropertyValue.java:38)
	at org.gradle.api.internal.tasks.DefaultTaskInputPropertySpec.validate(DefaultTaskInputPropertySpec.java:53)
	at org.gradle.api.internal.tasks.DefaultTaskInputs.validate(DefaultTaskInputs.java:168)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:47)
	... 24 more

-----------------------

The root cause of this failure is a bug in the pattern matching code in DefaultJvmVersionDetector::parseJavaVersionCommandOutput which was exposed by the additonal text added to the version string for JEP 322 (Time-Based Release Versioning).

    Matcher matcher = Pattern.compile("(?:java|openjdk) version \"(.+?)\"").matcher(versionStr);
    if (matcher.matches()) {
        return JavaVersion.toVersion(matcher.group(1));
    }

Prior to jdk-10+36, the first line of java -version looked like this:

java "10-ea"

but now it can have additional text after the "10-ea" like this:

java "10-ea" 2018-03-20

The pattern matching code in gradle will not find a match if there is anything at all on the line after the second quote.

One possible fix is to use matcher::find rather than matcher::matches.


### Your Environment
gradle 4.4 (also fails with 4.3)
jdk-10+36

--- a/subprojects/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/DefaultJvmVersionDetector.java
+++ b/subprojects/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/DefaultJvmVersionDetector.java
@@ -62,7 +62,7 @@ public class DefaultJvmVersionDetector implements JvmVersionDetector {
         try {
             String versionStr = reader.readLine();
             while (versionStr != null) {
-                Matcher matcher = Pattern.compile("(?:java|openjdk) version \"(.+?)\"").matcher(versionStr);
+                Matcher matcher = Pattern.compile("(?:java|openjdk) version \"(.+?)\"( \\d{4}-\\d{2}-\\d{2}( LTS)?)?").matcher(versionStr);
                 if (matcher.matches()) {
                     return JavaVersion.toVersion(matcher.group(1));
                 }

--- a/subprojects/jvm-services/src/test/groovy/org/gradle/internal/jvm/inspection/DefaultJvmVersionDetectorTest.groovy
+++ b/subprojects/jvm-services/src/test/groovy/org/gradle/internal/jvm/inspection/DefaultJvmVersionDetectorTest.groovy
@@ -60,6 +60,18 @@ Java HotSpot(TM) 64-Bit Server VM (build 25.91-b14, mixed mode)
         Java(TM) SE Runtime Environment (build 1.8.0_91-b14)
         Java HotSpot(TM) 64-Bit Server VM (build 25.91-b14, mixed mode)
         """                     | "1.8"
+        """java version "9.0.1"
+        Java(TM) SE Runtime Environment (build 9.0.1+11)
+        Java HotSpot(TM) 64-Bit Server VM (build 9.0.1+11, mixed mode)
+        """                     | "9.0.1"
+        """java version "10-ea"
+        Java(TM) SE Runtime Environment 18.3 (build 10-ea+35)
+        Java HotSpot(TM) 64-Bit Server VM 18.3 (build 10-ea+35, mixed mode)
+        """                     | "10"
+        """java version "10-ea" 2018-03-20
+        Java(TM) SE Runtime Environment 18.3 (build 10-ea+36)
+        Java HotSpot(TM) 64-Bit Server VM 18.3 (build 10-ea+36, mixed mode)
+        """                     | "10"
     }
 
     def "fails to parse version number"() {

